### PR TITLE
Nikita/leftparking

### PR DESCRIPTION
**Pull request by mishra-nikita**
This is still buggy. Right parking works, left parking does not. (2nd goal does not get generated because 1st goal is never reached. It keeps adjusting even after setting a high enough threshold)
Can review the stuff for now and use it for further development.

Up to date with kevin/integration + eva/left_turns_with_back_cam

Updated:
Line centerpoint will be detected(and used for comparison) only until first goal is published

Further improvement:
Add an if block for centerline midpoint calculation only during parking maneuver

### CODE REVIEW

File: autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py [hunk 1]
```diff
@@ -78,12 +89,35 @@ def detectLaneRGBD(self, color_img, depth_img):
...
+
+            transf = self.tf_buffer.lookup_transform(self.target_id, # target_frame_id
+                                        self.source_id, # source frame
+                                        rospy.Time(0), # get the tf at first available time
+                                        rospy.Duration(1.0)) # timeout after 1
+            xyz = PointStamped()
+            xyz.point.x = center_line_cloud[0,0]
+            xyz.point.y = center_line_cloud[0,1]
+            xyz.point.z = center_line_cloud[0,2]                        
+            line = do_transform_point(xyz,transf).point
```
Comment (SubramanianKrish) [inline autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py]: rename to whatever it means for fut. ref.

File: autovalet_system/src/autovalet_system.py [hunk 2]
```diff
@@ -91,13 +91,17 @@ def __init__(self,sim):
         self.husky_frame        = 'base_link'
         self.aruco_frame_name   = 'parking_spot' #'aruco_marker_frame' or 'parking_spot'
                             # (Needs to be same as what is set in aruco launcher)
-
+        # self.target_id          = 'base_link'
```
Comment (SubramanianKrish) [inline autovalet_system/src/autovalet_system.py]: Remove if not necessary

File: autovalet_system/src/autovalet_system.py [hunk 3]
```diff
@@ -91,13 +91,17 @@ def __init__(self,sim):
...
+        # self.source_id          = 'frontCamera_color_optical_frame'
+        
         self.parker = parking_spot(self.goal_topic,
                                    self.tag_topic,
                                    self.map_frame,
                                    self.husky_frame,
                                    self.aruco_frame_name,
-                                   debug=False)
+                                #    self.target_id,
```
Comment (SubramanianKrish) [inline autovalet_system/src/autovalet_system.py]: remove if not needed

File: autovalet_system/src/autovalet_system.py [hunk 4]
```diff
@@ -139,7 +143,9 @@ def registered_image_callback(self, color_msg, depth_msg):
         # if we're not in the PARK state AND the lane detector has been successfully initialized, detect the lane and publish
         if self.current_state != State.PARK and self.ld_init:
             # lane detection algo
-            _, self.ego_line = self.laneDetector.detectLaneRGBD(self.color_img, self.depth_img)
+            foo, self.ego_line, temp_parker_line = self.laneDetector.detectLaneRGBD(self.color_img, self.depth_img)
```
Comment (SubramanianKrish) [inline autovalet_system/src/autovalet_system.py]: rename to meaning names

**Review by SubramanianKrish — COMMENTED** [review]
Clean up based on comments

File: autovalet_system/src/autovalet_system.py [hunk 5]
```diff
@@ -139,7 +143,9 @@ def registered_image_callback(self, color_msg, depth_msg):
...
+            if temp_parker_line is not None:
```
Comment (kob51) [inline autovalet_system/src/autovalet_system.py]: rename temp_parker_line

File: autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py (hunk 1 above)
Comment (mishra-nikita) [reply autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py]: changed to centerline_midpt

Comment (mishra-nikita) [reply autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py]: and tmp_centerline_midpt

File: autovalet_system/src/autovalet_system.py (hunk 2 above)
Comment (mishra-nikita) [reply autovalet_system/src/autovalet_system.py]: removed

File: autovalet_system/src/autovalet_system.py (hunk 3 above)
Comment (mishra-nikita) [reply autovalet_system/src/autovalet_system.py]: removed

File: autovalet_system/src/autovalet_system.py (hunk 4 above)
Comment (mishra-nikita) [reply autovalet_system/src/autovalet_system.py]: done

File: autovalet_system/src/autovalet_system.py (hunk 5 above)
Comment (mishra-nikita) [reply autovalet_system/src/autovalet_system.py]: done

File: autovalet_husky/src/autovalet_husky/find_parking_spot.py [hunk 6]
```diff
@@ -39,6 +39,9 @@ def __init__(self, goal_topic, tag_topic, goal_frame_id, husky_frame_id, aruco_f
         self.aruco_frame_name = aruco_frame_name
         self.goal_frame_id    = goal_frame_id
         self.husky_frame_id   = husky_frame_id
+        self.first_goal_close = False
```
Comment (evamo0508) [inline autovalet_husky/src/autovalet_husky/find_parking_spot.py]: remove this line. I renamed it in my earlier MR and I think it slipped through your branch when you're rebasing.

File: autovalet_husky/src/autovalet_husky/find_parking_spot.py [hunk 7]
```diff
@@ -132,6 +149,7 @@ def generate_parking_goal(self, tf, pos, rot):
         waypoint published to movebase
         '''
         tag2waypoint = PoseStamped()
+        tag2waypoint.header.frame_id = 'parking_spot'
```
Comment (evamo0508) [inline autovalet_husky/src/autovalet_husky/find_parking_spot.py]: should be `self.aruco_frame_name`?

File: autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py [hunk 8]
```diff
@@ -12,13 +12,20 @@
 # python libs
 import numpy as np
 import cv2
+from cv_bridge import CvBridge, CvBridgeError
```
Comment (evamo0508) [inline autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py]: redundant. same as line 26.

File: autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py [hunk 9]
```diff
@@ -78,12 +89,35 @@ def detectLaneRGBD(self, color_img, depth_img):
             publishCloud(ego_line, self.camera.header.frame_id, self.egoLine_pub)
             # maybe publish center line alone too?
 
-            return lane_cloud, ego_line
+            ## Parking directions
+            if  center_line_coordinates.shape[0] % 2 == 0:
+                index = center_line_coordinates.shape[0] / 2 
+            else:
+                index = (center_line_coordinates.shape[0] + 1)/ 2
```
Comment (evamo0508) [inline autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py]: These could be refactored into one line code?
`index = int(round((center_line_coordinates.shape[0] + 1) / 2 ))`

File: autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py [hunk 10]
```diff
@@ -78,12 +89,35 @@ def detectLaneRGBD(self, color_img, depth_img):
...
+            center_line_coordinates= center_line_coordinates[index,:]
+            
+            if center_line_coordinates.shape[0] != 0:
+                center_line_cloud = self.line2cloud(depth_img, center_line_coordinates.reshape(1,-1))    # px3      
+
+            transf = self.tf_buffer.lookup_transform(self.target_id, # target_frame_id
+                                        self.source_id, # source frame
+                                        rospy.Time(0), # get the tf at first available time
+                                        rospy.Duration(1.0)) # timeout after 1
```
Comment (evamo0508) [inline autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py]: I'm being OCD but can you align the arguments plz 🙏

File: autovalet_system/launch/state_machine.launch [hunk 11]
```diff
@@ -30,7 +30,7 @@
 
         <!-- STATE MACHINE -->
         <!-- state machine node runs lane detection, goal generation, and parking subsystems, as well as overall system logic -->
-        <node name="STATE_MACHINE" pkg="autovalet_system" type="run_sm.sh" args="$(arg simulation)" />
+        <!-- <node name="STATE_MACHINE" pkg="autovalet_system" type="run_sm.sh" args="$(arg simulation)" /> -->
```
Comment (evamo0508) [inline autovalet_system/launch/state_machine.launch]: don't change this? we still want a not-debugging version on `master`.

File: autovalet_husky/src/autovalet_husky/find_parking_spot.py (hunk 6 above)
Comment (mishra-nikita) [reply autovalet_husky/src/autovalet_husky/find_parking_spot.py]: ok done

File: autovalet_husky/src/autovalet_husky/find_parking_spot.py (hunk 7 above)
Comment (mishra-nikita) [reply autovalet_husky/src/autovalet_husky/find_parking_spot.py]: great catch. Subbu had changed that for quick testing I think. And we forgot to revert the change.
Fixed it.

File: autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py (hunk 8 above)
Comment (mishra-nikita) [reply autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py]: oops done

File: autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py (hunk 10 above)
Comment (mishra-nikita) [reply autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py]: lol done

File: autovalet_system/launch/state_machine.launch (hunk 11 above)
Comment (mishra-nikita) [reply autovalet_system/launch/state_machine.launch]: yes i had it uncommented, just haven't pushed that change yet

File: autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py (hunk 9 above)
Comment (mishra-nikita) [reply autovalet_lane_detection/src/autovalet_lane_detection/LaneDetector.py]: perfect. done

**Pull request merged**